drivers/pci/controller/vmd.c

```diff
@@ -767,17 +831,239 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+		 * stored in PCI_MEMBAR1_OFFSET and PCI_MEMBAR2_OFFSET registers
+		 */
+		vmd_configure_membar(vmd, VMD_RESOURCE_MEMBAR_1, VMD_MEMBAR1, 0,
+				     pch_membar1_offset, NULL);
+		vmd_configure_membar(vmd, VMD_RESOURCE_MEMBAR_2, VMD_MEMBAR2,
+				     membar2_offset,
+				     pch_membar2_offset - membar2_offset, NULL);
+
+		vmd_configure_membar(vmd, VMD_RESOURCE_PCH_MEMBAR_1,
+				     VMD_MEMBAR1, pch_membar1_offset, 0,
+				     &vmd->resources[VMD_RESOURCE_MEMBAR_1]);
+		vmd_configure_membar(vmd, VMD_RESOURCE_PCH_MEMBAR_2,
+				     VMD_MEMBAR2,
+				     membar2_offset + pch_membar2_offset, 0,
+				     &vmd->resources[VMD_RESOURCE_MEMBAR_2]);
+	} else {
+		vmd_configure_membar(vmd, VMD_RESOURCE_MEMBAR_1, VMD_MEMBAR1, 0,
+				     0, NULL);
+		vmd_configure_membar(vmd, VMD_RESOURCE_MEMBAR_2, VMD_MEMBAR2,
+				     membar2_offset, 0, NULL);
+	}
+}
+
+static void vmd_bus_enumeration(struct pci_bus *bus, unsigned long features)
+{
+	struct pci_bus *child;
+	struct pci_dev *dev;
+	int ret;
+
+	vmd_acpi_begin();
+
+	pci_scan_child_bus(bus);
+
+	if (bus->primary == 0)
+		vmd_domain_reset(vmd_from_bus(bus));
+
+	/*
+	 * When Intel VMD is enabled, the OS does not discover the Root Ports
+	 * owned by Intel VMD within the MMCFG space. pci_reset_bus() applies
+	 * a reset to the parent of the PCI device supplied as argument. This
+	 * is why we pass a child device, so the reset can be triggered at
+	 * the Intel bridge level and propagated to all the children in the
+	 * hierarchy.
+	 */
+	list_for_each_entry(child, &bus->children, node) {
+		if (!list_empty(&child->devices)) {
+			dev = list_first_entry(&child->devices, struct pci_dev,
+					       bus_list);
+			ret = pci_reset_bus(dev);
+			if (ret)
+				pci_warn(dev, "can't reset device: %d\n", ret);
+
+			break;
+		}
+	}
+
+	pci_assign_unassigned_bus_resources(bus);
+
+	pci_walk_bus(bus, vmd_pm_enable_quirk, &features);
+
+	/*
+	 * VMD root buses are virtual and don't return true on pci_is_pcie()
+	 * and will fail pcie_bus_configure_settings() early. It can instead be
+	 * run on each of the real root ports.
+	 */
+	list_for_each_entry(child, &bus->children, node)
+		pcie_bus_configure_settings(child);
+
+	pci_bus_add_devices(bus);
+
+	vmd_acpi_end();
+}
+
+static int vmd_create_pch_bus(struct vmd_dev *vmd, struct pci_sysdata *sd,
+			      resource_size_t *offset)
+{
+	LIST_HEAD(resources_pch);
+
+	pci_add_resource(&resources_pch,
+			 &vmd->resources[VMD_RESOURCE_PCH_CFGBAR]);
+	pci_add_resource_offset(&resources_pch,
+				&vmd->resources[VMD_RESOURCE_PCH_MEMBAR_1],
+				offset[0]);
+	pci_add_resource_offset(&resources_pch,
+				&vmd->resources[VMD_RESOURCE_PCH_MEMBAR_2],
+				offset[1]);
+
+	vmd->bus[VMD_ROOTBUS_1] =
+		pci_create_root_bus(&vmd->dev->dev, vmd->busn_start[VMD_ROOTBUS_1],
+					&vmd_ops, sd, &resources_pch);
+
+	if (!vmd->bus[VMD_ROOTBUS_1]) {
+		pci_free_resource_list(&resources_pch);
+		vmd_remove_irq_domain(vmd);
```

Comment:
Should we? We didn't register PCH MSI yet.

Reply:
Shoudn't we call `pci_stop_root_bus` and `pci_remove_root_bus`?

Reply:
correct, irq domain cannot be removed here, I have added stopping and removig PCH bus instead.